[Feature] Image Processing tab

devilismyfriend opened the pull request:
Hi

UI:
<img width="1034" alt="image" src="https://user-images.githubusercontent.com/87043616/187788890-dd258ceb-d3b4-428a-bb18-982ae1fafe93.png">


Here's a refactoring of the way the project handles upscaling, changes:

-Removal of Fix Faces and Upscale options from the generating interface (commented out currently in case we want to keep it)
-Addition of a new tab called Image Lab
-Addition of Single image and Batch processors 
-Cleaner Reimplementation of GoBig 
-Addition of Latent Diffusion Super Resolution upscaler
-Multiple processors in the same run
-Provides a location and base to add more processors like Animation generation or addition of future upscalers.

Future additions:
-Copy generation settings from generation interface to the goBig upscale settings for convenience (not big on JS, will take some time unless someone wants to tackle it)
-Batch process GoBig settings extractor from current file, using metadata or filename.
 
To use LDSR users will need to clone this custom repo https://github.com/devilismyfriend/latent-diffusion to ./src

LDSR was originally an example collab/cog project, I added a batch script to use it locally and also compiled into a module that can be loaded as model for this project.

After cloning users can use the download_model.bat to download and place the latest model files, locations follow previous established workflows and they should be placed under the ./src/latent-diffusion/experiments/pretrained_models.

Comparison of current upscalers:

Original 448x640
![image](https://user-images.githubusercontent.com/87043616/187347736-cc5ae769-1a74-4faa-8b6f-a35ccc62e646.png)


RealESRGAN, Fast, 4X, loss of details
![image](https://user-images.githubusercontent.com/87043616/187347816-a31ea87b-608e-461d-8976-5874fa0296d7.png)

GoBig, RealESRGAN 2X -> Sliced and SD diffused to add details, takes prompt/seed etc that will hopefully be pushable from the other tabs
![image](https://user-images.githubusercontent.com/87043616/187347860-62c0d9f0-4a9a-40fd-941d-0da7bc49ba03.png)


Latent Diffusion, Slow, 4X, doesn't alter the image like GoBig, produces some grain but is sharper and more accurate then RealESRGAN while providing better resolution then GoBig.
![image](https://user-images.githubusercontent.com/87043616/187347897-ec0bc99a-9cbf-4345-8b1d-8400faf90b88.png)

GoLatent, Slowest, 8X, combines the GoBig workflow with the Latent Diffusion upscaling method to add details and provide a clearer image through Latent Diffusion.
![image](https://i.imgur.com/sLvTLbI.jpg)

SkylarKelty commented:
This is brilliant work, I was hoping someone would get GoBIG in but the extra GoLatent option is awsome
One thing: I think `def imgproc` is returning the original instead of the modified image, so the web ui "result" section ends up being the original.

devilismyfriend commented:
> With this can realersgan be used just at 2x without gobig? if not that would be a nice add

Yes, you can select a 2x model

devilismyfriend commented:
> This is brilliant work, I was hoping someone would get GoBIG in but the extra GoLatent option is awsome One thing: I think `def imgproc` is returning the original instead of the modified image, so the web ui "result" section ends up being the original.

lol, thx for the heads up, guess i was tired last night haha, will fix in the next hour

altryne commented:
Dang you drop this on us while @hlky went to sleep? 
Great job, I'll start a review

nderscore commented on frontend/frontend.py:
```diff
@@ -293,7 +293,122 @@ def img2img_submit_params():
+                                    imgproc_folder = gr.File(label="Batch Process", file_count="multiple",source="upload", interactive=True, type="file")
+                            with gr.Row():
+                                imgproc_btn = gr.Button("Process", variant="primary")
+                            gr.HTML("""
```
Minor: this (and the other text sections) would be a lot easier to maintain as markdown instead of html markup. Centering the text also makes it harder to read.

devilismyfriend replied:
Well first of all I agree that it should live in the wiki, the problem is that there's a large subset of peeps that are getting into AI generated stuff while having no knowledge about more advanced topics, I can't expect people to know what GoBig is or Latent Diffusion, which is why I included some small info there, plus it pads out the UI to make it look better when the upscale options are open lol.

In regards to HTML vs Markdown, I actually had it as Markdown before, but decided to use HTML instead so I can use the same style as the message at the bottom of the UI, may be it should be defined outside of the loop tho.

nderscore replied:
The message in the footer is long. This one is multiple paragraphs, and it's nested inside of a column in a container. It would really be easier to read if it was left-aligned normally :sweat_smile: (that space on the left side should be reserved for the controls/settings probably?)

devilismyfriend replied:
Amm hard for me to tell, on my screen it looks fine, but I evidently didn't try it on others, what device are you viewing it on?

grexzen commented:
Tried it, and seems to load just fine, but still says this on UI: "Please download LDSR to activate more upscale features, instructions are available at the [Github](https://github.com/hlky/stable-diffusion-webui)".

Console says successfully installed, and "LatentDiffusion: Running in eps-prediction mode".

What is eps-prediction mode? Do I need to do anything else besides following instructions above? 

Thanks!

devilismyfriend commented:
> Tried it, and seems to load just fine, but still says this on UI: "Please download LDSR to activate more upscale features, instructions are available at the [Github](https://github.com/hlky/stable-diffusion-webui)".
> 
> Console says successfully installed, and "LatentDiffusion: Running in eps-prediction mode".
> 
> What is eps-prediction mode? Do I need to do anything else besides following instructions above?
> 
> Thanks!

To use LDSR users will need to clone this custom repo https://github.com/devilismyfriend/latent-diffusion to ./src

So basically, clone that repo, into your src folder, go into it and use the download_model.bat, it should download the required model and config file and place them under src/latent-diffusion/experiments/pretrained_models/, you should have model.ckpt and project.yml there

grexzen commented:
Thank you. Did all that the first time, and just repeated the process to make sure, still nope. Files are there. 

And console says "Successfully installed GFPGAN clip facexlib k-diffusion latent-diffusion realesrgan taming-transformers". So it is finding the model.

Also, heads up there is no Real x2 on the other module folder, so the console throws "RealESRGAN_x2plus.pth not found at path ./src/realesrgan\experiments/pretrained_models\RealESRGAN_x2plus.pth"

I tried finding it, but can't see it. https://github.com/xinntao/Real-ESRGAN

nderscore commented:
`RealESRGAN` still works.

`GoBig` mode crashes with this trace: (changing to different models doesn't seem to make a difference either)
```
    result = context.run(func, *args)
  File "scripts/webui.py", line 1721, in imgproc
    image = processGoBig(image)
  File "scripts/webui.py", line 1478, in processGoBig
    sampler = KDiffusionSampler(model,'lms')
NameError: name 'model' is not defined
```

`Latent Diffusion SR` mode crashes for me with this trace:
```
  File "scripts/webui.py", line 1740, in imgproc
    image = processLDSR(image)
  File "scripts/webui.py", line 1659, in processLDSR
    result = LDSR.superResolution(image)
NameError: name 'LDSR' is not defined
```

`GoLatent` worked! The `superResolution` stage (I think? it took 100 steps, regardless of my settings) took 8 minutes on my RX 6900 XT (might be being bottle-necked by my old CPU?) but I got a `GoLatent` upscale :sunglasses:

devilismyfriend commented:
> 

Thanks for testing it out, I submitted changes, see if they fix your issues, I added forced load checks to make sure they're loaded prior to the operation, I need to write a better load/unload module to properly control the models but for now I hope this will fix it

devilismyfriend commented:
> `RealESRGAN` still works.
> 
> `GoBig` mode crashes with this trace: (changing to different samplers doesn't seem to make a difference either)
> 
> ```
>     result = context.run(func, *args)
>   File "scripts/webui.py", line 1721, in imgproc
>     image = processGoBig(image)
>   File "scripts/webui.py", line 1478, in processGoBig
>     sampler = KDiffusionSampler(model,'lms')
> NameError: name 'model' is not defined
> ```
> 
> `Latent Diffusion SR` mode crashes for me with this trace:
> 
> ```
>   File "scripts/webui.py", line 1740, in imgproc
>     image = processLDSR(image)
>   File "scripts/webui.py", line 1659, in processLDSR
>     result = LDSR.superResolution(image)
> NameError: name 'LDSR' is not defined
> ```
> 
> `GoLatent` worked! The `superResolution` stage (I think? it took 100 steps, regardless of my settings) took 8 minutes on my RX 6900 XT (might be being bottle-necked by my old CPU?) but I got a `GoLatent` upscale 😎

Thanks for testing it out, I submitted changes, see if they fix your issues, I added forced load checks to make sure they're loaded prior to the operation, I need to write a better load/unload module to properly control the models but for now I hope this will fix it

devilismyfriend commented:
> Thank you. Did all that the first time, and just repeated the process to make sure, still nope. Files are there.
> 
> And console says "Successfully installed GFPGAN clip facexlib k-diffusion latent-diffusion realesrgan taming-transformers". So it is finding the model.
> 
> Also, heads up there is no Real x2 on the other module folder, so the console throws "RealESRGAN_x2plus.pth not found at path ./src/realesrgan\experiments/pretrained_models\RealESRGAN_x2plus.pth"
> 
> I tried finding it, but can't see it. https://github.com/xinntao/Real-ESRGAN

"Successfully installed GFPGAN clip facexlib k-diffusion latent-diffusion realesrgan taming-transformers" is what you'd get if you're trying to activate the enviroment?, did you clone the repo I linked, went into it and ran the bat file to download the models? if the files exists in this path: "\src\latent-diffusion\experiments\pretrained_models\model.ckpt" it should load.

In regards to the 2x mode, I just posted a fix for it, but yeah, there's no real 2x model file, I just 4x and resize it after the fact.

devilismyfriend commented:
> > Let me know if this fixed your issue for now :)
> 
> So far so good with the latest commits 👍
> 
> `GoBig` works. `Latent Diffusion SR` works. I'm running `GoLatent` again right now and based on how it's gone so far, I think its same to assume it's still working.

Nice! Thanks :D

devilismyfriend commented:
> Thank you. Did all that the first time, and just repeated the process to make sure, still nope. Files are there.
> 
> And console says "Successfully installed GFPGAN clip facexlib k-diffusion latent-diffusion realesrgan taming-transformers". So it is finding the model.
> 
> Also, heads up there is no Real x2 on the other module folder, so the console throws "RealESRGAN_x2plus.pth not found at path ./src/realesrgan\experiments/pretrained_models\RealESRGAN_x2plus.pth"
> 
> I tried finding it, but can't see it. https://github.com/xinntao/Real-ESRGAN

OK I looked into it, the webui.cmd which I believe you're using to launch the UI actually runs /scripts/webui.py, so if you want it to work you should copy the webui.py from this PR to the scripts folder and replace the webui.py there.

grexzen commented:
![Screenshot 2022-08-31 180048](https://user-images.githubusercontent.com/30579087/187812774-33ebb1d8-9bc2-466e-ab5f-3be66612cfed.jpg)

Yep, updated the main repo, cloned your repo, placed on scr, ran bat, verified model and yaml file are there, then replaced scripts/webui and frontend/frontend files with yours. 

Pic shows the model is in the folder, but code says nope. 

Do I need to do the Step 2a optional from the guide first? https://rentry.org/GUItard Maybe that's why is not working.

grexzen commented:
![Screenshot 2022-08-31 180048](https://user-images.githubusercontent.com/30579087/187813945-4401551e-247b-4202-b359-0bb57abcf239.jpg)

And to clarify, I do see the UI changes for sure.

devilismyfriend commented:
> ![Screenshot 2022-08-31 180048](https://user-images.githubusercontent.com/30579087/187812774-33ebb1d8-9bc2-466e-ab5f-3be66612cfed.jpg)
> 
> Yep, updated the main repo, cloned your repo, placed on scr, ran bat, verified model and yaml file are there, then replaced scripts/webui and frontend/frontend files with yours.
> 
> Pic shows the model is in the folder, but code says nope.
> 
> Do I need to do the Step 2a optional from the guide first? https://rentry.org/GUItard Maybe that's why is not working.

ammm are you launching the webui.cmd from the stable-diffusion-folder?, it's obviously not finding the path but since it does exist I can only think it's a working directory issue.

grexzen commented:
I also found the real x2 model at the bottom of this page.

https://upscale.wiki/wiki/Model_Database#Real-ESRGAN_Compact_Custom_Models 

https://github.com/xinntao/Real-ESRGAN/releases/download/v0.2.1/RealESRGAN_x2plus.pth

devilismyfriend commented:
> I also found the real x2 model at the bottom of this page.
> 
> https://upscale.wiki/wiki/Model_Database#Real-ESRGAN_Compact_Custom_Models
> 
> https://github.com/xinntao/Real-ESRGAN/releases/download/v0.2.1/RealESRGAN_x2plus.pth

cool

You can also try copying the latent-diffusion folder, renaming it LDSR and putting it in the main directory

ChrisTopherTa54321 commented:
> ![Screenshot 2022-08-31 180048](https://user-images.githubusercontent.com/30579087/187812774-33ebb1d8-9bc2-466e-ab5f-3be66612cfed.jpg)
> 
> Yep, updated the main repo, cloned your repo, placed on scr, ran bat, verified model and yaml file are there, then replaced scripts/webui and frontend/frontend files with yours.
> 
> Pic shows the model is in the folder, but code says nope.

Your Explorer window shows the path src/latent-diffusion, your error message shows the path src/LDSR...

JoshuaKimsey commented:
> To use LDSR users will need to clone this custom repo https://github.com/devilismyfriend/latent-diffusion to ./src

> After cloning users can use the download_model.bat to download and place the latest model files, locations follow previous established workflows and they should be placed under the ./src/latent-diffusion/experiments/pretrained_models.

Heh, if this is how the final version stays when its merged, I definitely know what I'll be updating my Linuxx Bash script to be doing! This is absolutely amazing work!

devilismyfriend commented:
Cleaned up the entire model loading situation and removed global variables (for imgproc), added a new model loader/unloader function and now each model can be loaded or unloaded before the process properly.

Tollanador commented:
> ![Screenshot 2022-08-31 180048](https://user-images.githubusercontent.com/30579087/187812774-33ebb1d8-9bc2-466e-ab5f-3be66612cfed.jpg)
> 
> Yep, updated the main repo, cloned your repo, placed on scr, ran bat, verified model and yaml file are there, then replaced scripts/webui and frontend/frontend files with yours.
> 
> Pic shows the model is in the folder, but code says nope.
> 
> Do I need to do the Step 2a optional from the guide first? https://rentry.org/GUItard Maybe that's why is not working.

There is a Forward slash after LDSR when it should be Backslash, not sure if that is the reason for your error.

I've been playing with this too. It's great!

devilismyfriend commented:
> > ![Screenshot 2022-08-31 180048](https://user-images.githubusercontent.com/30579087/187812774-33ebb1d8-9bc2-466e-ab5f-3be66612cfed.jpg)
> > Yep, updated the main repo, cloned your repo, placed on scr, ran bat, verified model and yaml file are there, then replaced scripts/webui and frontend/frontend files with yours.
> > Pic shows the model is in the folder, but code says nope.
> > Do I need to do the Step 2a optional from the guide first? https://rentry.org/GUItard Maybe that's why is not working.
> 
> There is a Forward slash after LDSR when it should be Backslash, not sure if that is the reason for your error.
> 
> I've been playing with this too. It's great!

I don't think it matters, the fact ./LDSR is there in the first place is because the webui isn't finding ./src or the following folders (it defaults to that value as do the other upscalers)

Problem is by now who the hell knows where people are getting their setups from with the countless available packages, some are using the webui.cmd which doesn't even run the main webui.py lol, hard to tell what the issue is.

Tollanador commented:
I installed mine is if I would if it was the main repo, using my Colab, which doesn't use miniconda environment. With the addition of the Latent Diffusion model, I had to switch to the High Memory Runtime .. worth it for this awesome implementation feature!

devilismyfriend commented:
> I installed mine is if I would if it was the main repo, using my Colab, which doesn't use miniconda environment. With the addition of the Latent Diffusion model, I had to switch to the High Memory Runtime .. worth it for this awesome implementation feature!

Yeah it's quite simple, but for peeps who are venturing out of their comfort zone in order to play with the AI, it's a bit cumbersome to get going currently.

Tollanador commented:
Yeah, it's at this point I'm glad I chose to not use the Miniconda approach in Colab and just run a cell that has all the installing, setting up, etc.
I then have a few cells that can switch between Main, Development, This version and the Old GoBig version.

hlky commented:
Tests ok, little delayed showing the output due to reloading the models.

TODO: grid showing in output gallery with image count = 1, this was previously fixed.

TODO: document new upscalers

devilismyfriend commented:
> If running in optimized mode modelCS and modelFS is not made available later in the script for gobig I need to add
> 
> ```
> global modelCS 
> global modelFS 
> ```
> 
> for it to work

thanks, fixed it.

ooofest commented:
This is a very nice feature, but I see no reason to remove the convenience of GFPGAN and ESRGAN remaining options in the Text-to-Image tab, because it is far more convenient for batch txt2img processing and even saving manual labor for single batch runs.

I've been trying the new workflow after removal of those options from the Text-to-Image tab and it's already frustrating.  I have seen the WebUI effort as not only bringing cool new/upgraded features to the use of SD, but also enabling more efficient processing.  This latest update has removed the latter, IMHO.

We can have features in both tabs - they need not be mutally exclusive.

devilismyfriend commented:
> This is a very nice feature, but I see no reason to remove the convenience of GFPGAN and ESRGAN remaining options in the Text-to-Image tab, because it is far more convenient for batch txt2img processing and even saving manual labor for single batch runs.
> 
> I've been trying the new workflow after removal of those options from the Text-to-Image tab and it's already frustrating. I have seen the WebUI effort as not only bringing cool new/upgraded features to the use of SD, but also enabling more efficient processing. This latest update has removed the latter, IMHO.
> 
> We can have features in both tabs - they need not be mutally exclusive.

The idea is to separate post processing and generation, that being said I hear you, it might not be the worst idea having quick access to common procedures like those, I have another PR open that will add a send to lab button to quickly move between the generation and also bring back the advanced options if you'd like to stay in that tab.

I feel this still needs to be discussed to some extent but no reason to take functionality away if people deem it necessary

ooofest commented:
To that point, it would seem an appropriate experiment to resurrect the face/upscale options on the txt2img tab and gain feedback on whether that is not useful or even confusing to users.  I'm not sure the automatic/batch use cases were part of the initial consideration when pulling those options from txt2img, that is - perhaps I'm wrong, though.

And, to be clear, I feel the old options do not need to be added back as they were to achieve the same effect for end users: e.g., the txt2img tab options to modify face/upscale might trigger an automated batch option to feed the Image Lab functions (if possible) after each txt2img (or img2img) output has been generated, as one idea.  This could take advantage of the new Image Lab options and reuse them in automated/batch cases, for example.

Something to consider.

Maergoth commented:
> ![Screenshot 2022-08-31 180048](https://user-images.githubusercontent.com/30579087/187813945-4401551e-247b-4202-b359-0bb57abcf239.jpg)
> 
> And to clarify, I do see the UI changes for sure.

Did you get this resolved? I have the same issue. Followed the steps, no button. 

![image](https://user-images.githubusercontent.com/93102337/187990845-53faf1c7-4408-485d-9ad5-c15420fe87f0.png)

altryne commented:
>To that point, it would seem an appropriate experiment to resurrect the face/upscale options on the txt2img tab and gain feedback on whether that is not useful or even confusing to users. I'm not sure the automatic/batch use cases were part of the initial consideration when pulling those options from txt2img, that is - perhaps I'm wrong, though.

I want to restore them, to me it was useful, I think to others as well. 

@devilismyfriend is there links to the new models somewhere? this needs to documentation (and we have an auto downloader in colab, I want to make sure colab users get all the new features

devilismyfriend commented:
> > To that point, it would seem an appropriate experiment to resurrect the face/upscale options on the txt2img tab and gain feedback on whether that is not useful or even confusing to users. I'm not sure the automatic/batch use cases were part of the initial consideration when pulling those options from txt2img, that is - perhaps I'm wrong, though.
> 
> I want to restore them, to me it was useful, I think to others as well.
> 
> @devilismyfriend is there links to the new models somewhere? this needs to documentation (and we have an auto downloader in colab, I want to make sure colab users get all the new features

I've got an open PR that brings them back along with more stuff, including an updated readme with links and instructions

geowjm commented:
hi, i got this when hitting "process" in any option : 
```
G:\envs\ldm\lib\site-packages\anyio\_backends\_asyncio.py", line 867, in run
    result = context.run(func, *args)
TypeError: imgproc() missing 3 required positional arguments: 'imgproc_ldsr_steps', 'imgproc_ldsr_pre_downSample', and 'imgproc_ldsr_post_downSample
```

devilismyfriend commented:
> hi, i got this when hitting "process" in any option :
> 
> ```
> G:\envs\ldm\lib\site-packages\anyio\_backends\_asyncio.py", line 867, in run
>     result = context.run(func, *args)
> TypeError: imgproc() missing 3 required positional arguments: 'imgproc_ldsr_steps', 'imgproc_ldsr_pre_downSample', and 'imgproc_ldsr_post_downSample
> ```

those changes arent merged yet, you took it from my fork?
